Latest, with ch4 (#18)
riscv-non-isa/riscv-security-model: specification/src/chapter2.adoc

* Proprietary models
* Secret algorithms

> ==== Trusted compute base

Reviewer: ```suggestion
==== Trusted Computing Base (TCB)
```

> The _trusted compute base (TCB)_ of a component is the minimum amount of

Reviewer: ```suggestion
The _Trusted Computing Base (TCB)_ of a component is the minimum amount of
```

software and hardware that needs to be trusted by that component. 

The TCB of a component typically includes other software. For example:

* An operating system

> * A hypervisor and a guest operating system

Reviewer: ```suggestion
* A hypervisor and a guest operating system
* A TEE security manager
```

* Hosting services such as orchestration and server provisioning software

==== Root of trust

> The minimum amount of software that always has to be trusted on a system is its _root of trust (RoT)_. A RoT includes fundamental security services, for example:

Reviewer: ```suggestion
The minimum amount of software that always has to be trusted on a system is its _root of trust (RoT)_. A RoT supports fundamental security services, for example:
```

* A dedicated trusted subsystem (HW RoT), supporting a FW RoT

Using a HW RoT moves critical functions and assets off a Hart to a dedicated
trusted subsystem, which can provide stronger protection against physical and logical attack than a complex Hart.

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| CAT_NNN
| A secure system SHOULD implement a HW RoT

> 

Reviewer 2: Should it state - the penultimate RoT in a secure system should be a HW RoT?

Author: There is a not already that a system can have multiple trust chains with their own RoT (secondary RoT), but the HW RoT is the primary RoT on a system. I moved that note earlier in the section and clarified.

| CAT_NNN

> | A FW RoT SHOULD be the only software in machine mode (M).

Reviewer: ```suggestion
| Software in machine mode (M) SHOULD be measured by the HW RoT e.g. FW RoT, Security services for TEE etc. This software SHOULD allow for updates under control of the HW RoT.
```

Author: This is too much information at this point, with concepts that have not been introduced yet. There are separate sections spelling out rules for software authorization, updates, etc.

Also, not all systems will have, and not all systems will need, a HW RoT. Obviously all CoVE systems will, and all GP TEE systems too. But not ALL RISC-V systems will need one.

I have reworked this section a bit to clarify.

trust. The HW RoT governs the security platform itself and acts as a primary
root of trust on the system.

==== Isolation

Assets can be protected by _isolation_. Isolation reduces dependencies between components, and reduces the amount of software that needs to be trusted. 

Isolation protects _resources_:

* Memory and memory mapped devices
* _Execution state_, including Hart register state

Examples of isolation mechanisms include:

* Privilege based isolation +
More privileged software is able to enforce security guarantees for less
privileged software.
* Physical memory isolation +
More privileged software controls memory access for less privileged software.
* Domain isolation +
Software in one domain cannot access or modify resources assigned to a different domain (without consent), regardless of privilege level. +
(Higher privileged software in one domain cannot access resources assigned to a lower privilege level in a different domain)
* Virtualization +
Virtualization creates and manages _virtual resources_ - compute, memory,
devices - independent of actual physical hardware. A system, or individual

> domains, can be virtualized.

Reviewer: should we provide references for each of these to the priv ISA or other specs?

| CAT_NNN
| Invasive subsystems MUST be controlled, or moderated, by a RoT.

|===

==== Event counters

Event counters are commonly used for performance management and resource allocation. 

However, they can also pose a security risk. For example, one workload measuring an operation in a different workload, or an operation by higher privilege software, could be able to reveal assets used in those operations.

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| CAT_NNN
| Lower privilege software MUST NOT be able to measure higher privilege software.

| CAT_NNN

> | Software in one domain MUST NOT be able to measure software in a different domain, without consent.

Reviewer 2: monitor instead of measure?

Reviewer 3: "monitor" in this case would be too broad of a word to use in this context as it takes into account side-channels. Would "architecturally monitor" be too verbose?

Reviewer: I think better to be specific - how about - 

Software in one domain MUST NOT be able to monitor software in a different domain using architectural performance monitoring events/counters, without consent.

Author: I changed to "Monitor" throughout the document.

These requirements are all in the context of Event counting, or architectural platform QoS. So taken in context of the sections they are in they should be clear enough without repeating context info everywhere?

Author: I did the same update in the surrounding text as well.

Server platforms can provide _platform quality of service (QoS)_ features, consisting of Hart and system hardware and firmware aimed at managing access to shared physical resources across workloads, minimizing contention. For example:

* Memory bandwidth management
* Cache allocation policies across workloads, including workload prioritization
* Hart allocation policies across workloads

These types of features rely on measuring resource utilization of workloads, similar to event counters, and optimizing resource allocation policies.

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| CAT_NNN
| Lower privilege software MUST NOT be able to measure higher privilege software.

| CAT_NNN
| Software in one domains MUST NOT be able to measure software in a different domain, without consent.

|===

==== Denial of service

The RISC-V security model is primarily concerned with protection of assets. 

_Denial of service (DoS)_ is out of scope of this specification. For example, a hosting environment is free to apply their own resource allocation policy to any workloads. Including denying service. This applies in the same way to confidential workloads.

> 

Reviewer 2: I think there may still want to be a statement that lower privileged SW should not be able to cause DoS on the resource manager - even if the resource manager is outside the TCB. For example, while a CVM having the ability to block all interrupts may be a security concern from an infrastructure protection perspective to the service provider.

Reviewer: +1 we should generally state that a "tenant" which may be a process/container/VM (ordinary VM or TVM) MUST NOT be able to launch a denial of service against the hosting platform (OS/VMM/orchestration).

Author: Thanks, good point. I rephrased the section and added a formal requirement.

Reviewer: nit - must MUST be capitalized :-)

Author: Not a normal sentence it doesn't ;-)

I changed the wording.

* Retrieving residual confidential memory after a system reset
a| * Implement robust power management
* Implement cryptographic memory protection with at least boot freshness

> * Code hardening (for example, see https://www.riscure.com/publication/practical-steps-evaluate-protect-secure-boot-implementations-embedded-devices/[Riscure white paper])

Reviewer 2: May want to move references to bibliography. This reference is not in public domain - should we replace with a public domain reference?

Author: It is public domain, they just ask for an email address. It is a commonly used reference for hardened boot, at least in IOT/edge use cases and many mobile clients, as Riscure also support several certification programmes.

If it is controversial then I can remove the reference. I just thought it easier to just link to that than repeat a lot of low-level technical detail. Happy to take suggestions for other references as well.

* Retrieving residual confidential memory after a system reset
a| * Implement robust power management
* Implement cryptographic memory protection with at least boot freshness

> * Code hardening (for example, see https://www.riscure.com/publication/practical-steps-evaluate-protect-secure-boot-implementations-embedded-devices/[Riscure white paper])

Reviewer 2: For references can include:
NIST SP 800-193 “Platform Firmware Resiliency Guidelines”
NIST SP 800-57 “Recommendation for Key Management”
NIST SP 800-147 “BIOS Protection Guidelines”
NIST SP 800-147B “BIOS Protection Guidelines for Servers”

A security lifecycle reflects the trustworthiness of a system during its
lifetime and reflects the lifecycle state of hardware provisioned assets.

It can be extended as indicated below to cover additional security provisioning steps such as device onboarding, device activation, user management, and RMA processes. These are use case or ecosystem specific and out of scope of this
specification.

For the purpose of this specification, _revealing debug_ includes any HW or FW debug capability which 

* Could break security guarantees or could expose assets 

> * Is not part of an attested trust contract with a reliant party 

Reviewer: ```suggestion
* Is not part of an attested trust contract with a relying party 
```

Examples of revealing debug include revealing logging, external debug or boundary scans, dedicated debug builds of software components, or enabling self-hosted debug for a component.

> Depending on use case, an attested software component can include debug capabilities managed through an ecosystem defined governance process - _trusted debug_. For example, self-hosted debug enabled following an ecosystem specific authorization process. In this case the debug capability, and the associated governance, is part of the trust contract with a reliant party.

Reviewer: ```suggestion
Depending on the use case, an attested software component can include debug capabilities managed through an ecosystem defined governance process - _trusted debug_. For example, self-hosted debug enabled following an ecosystem specific authorization process. In this case the debug capability, and the associated governance, is part of the trust contract with a relying party.
```

| Derived assets MUST only be available if a component is in secured state.
|===

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| CAT_NNN
| Hardware provisioned assets MUST only be accessible while the system is in

> secured state, or a recoverable debug state.

Reviewer: ```suggestion
secured state, or a recoverable debug state (with the recoverable debug state in attestation evidence).
```

| CAT_NNN
| A security platform attestation MUST be signed by a HW RoT, if present, or by

> a FW RoT

Reviewer: ```suggestion
a FW RoT (which has in its direct TCB a HW RoT)
```

Author: See comment above about HW RoT. We can't assume that every RISC-V system will have one.

Reviewer: how about adding - "a FW RoT (bootstrapped by a HW RoT if available on the system)"

| A layered attestation MAY be signed by lower privileged software, itself
attested by a security platform attestation

| CAT_NNN 

> a| Layered attestations MUST be cryptographically bound such that a reliant party can determine that they 

Reviewer: ```suggestion
a| Layered attestations MUST be cryptographically bound such that a relying party can determine that they 
```

| Requirement

| CAT_NNN

> | A system in secured or recoverable debug states MUST only load authorized software.

Reviewer: ```suggestion
| A system in secured or recoverable debug states MUST only load cryptographically authenticated software.
```

Author: The terms "authorized software" "measured software" and "verified software" are defined in the ecosystem chapter. I don't think we need to introduce new terms here.

| CAT_NNN

> | A system in secure provisioning state SHOULD only load authorized software.

Reviewer: ```suggestion
| A system in secure provisioning state SHOULD only load cryptographically authenticated software.
```

Author: The terms "authorized software" "measured software" and "verified software" are defined in the ecosystem chapter. I don't think we need to introduce new terms here.

| Requirement

| CAT_NNN
| A system MUST only use authorizations from trusted authority.
|===

> * Direct verification requires a signed image authorization from a trusted

Reviewer: ```suggestion
* Direct verification requires a signed image from a trusted
```

Author: The terms "authorized software" "measured software" and "verified software" are defined in the ecosystem chapter. I don't think we need to introduce new terms here.

Examples of authorizations in this case are in the sentence immediately following.

Isolation can also extend to other features, such as interrupts and debug.

==== Sealing

> Sealing is the process of protecting confidential assets on a system.

Reviewer: ```suggestion
Sealing is the process of protecting confidential assets on a system, typically by cryptographically binding sealing keys to the measured state of the workload that owns the assets.
```

Author: Different forms of sealing are defined next in this section.

Reviewer: that maybe but this introductory definition seems very incomplete.

Author: I see the problem now, this section is a little out of kilter with the rest. I have fixed it, and added the measurement based use case as well (already had that elsewhere).

| Requirement

| CAT_NNN
| Local sealed assets MUST only be possible to unseal on the same physical instance of a system that they were sealed on.

|===

> For example, using sealing keys derived from a hardware provisioned _hardware unique key (HUK)_.

Reviewer: ```suggestion
For example, using sealing keys derived from a silicon-vendor provisioned _hardware unique key (HUK)_.
```

Author: "Hardware provisioned assets" are used throughout this document. This specification does not define a specific provisioning model. Depending on use case they can be provisioned at any stage in a supply chain, not necessarily by a silicon vendor.

Reviewer: "hardware provisioned" sounds odd per se - since HW does not provision itself - I agree that the provisioning is done either by the silicon manufacturer, or some entity downstream in the supply chain - hence was suggesting some specific actor that does the provisioning. 

we could state it as : For example, using sealing keys derived from a silicon-vendor-(or other downstream entity)-provisioned _hardware unique key (HUK)_.

Author: I have used "hardware provisioned assets" throughout the document. Just realised it is not formally defined anywhere. Added a formal definition in the "Assets" section in the reference model (chapter 2) ("immutable asset provisioned in hardware by...").
